## Enron Data POI Classifier 
### Jo Anna Capp

In the 1990's Enron Corporation was one of the largest.....


In [1]:
#set working directory
import os
os.chdir('D:/Documents/Udacity/IntroMachineLearning/ud120projectsmaster/ud120projectsmaster/UdacityP5')

In [70]:
#import all packages and modules here
import sys
import pickle
sys.path.append("../tools/")
import pandas
import numpy
import seaborn as sns
import matplotlib.pyplot as plt
import sklearn
from sklearn import preprocessing
from ggplot import *
from sklearn.preprocessing import MinMaxScaler
from sklearn import cross_validation
from sklearn.cross_validation import train_test_split
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import SelectPercentile, f_classif
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score

from feature_format import featureFormat, targetFeatureSplit
from tester import dump_classifier_and_data

In [3]:
features_list = ['poi']

### Load the dictionary containing the dataset
with open("final_project_dataset.pkl", "r") as data_file:
    data_dict = pickle.load(data_file)

Lets look at the structure of the dataset and check for missing values.

In [4]:
#total individuals
print "There are ", len(data_dict.keys()), "executives of interest in the Enron dataset"
#number of pois
num_poi = 0
for dic in data_dict.values():
    if dic['poi'] == 1: 
        num_poi += 1
print "There are ", num_poi, "identified persons of interest within the dataset"
print "Data Dictionary Keys:"
print(data_dict.keys())
#data dictionary format
print "A typical key:value list: ", data_dict["SKILLING JEFFREY K"]


There are  146 executives of interest in the Enron dataset
There are  18 identified persons of interest within the dataset
Data Dictionary Keys:
['METTS MARK', 'BAXTER JOHN C', 'ELLIOTT STEVEN', 'CORDES WILLIAM R', 'HANNON KEVIN P', 'MORDAUNT KRISTINA M', 'MEYER ROCKFORD G', 'MCMAHON JEFFREY', 'HORTON STANLEY C', 'PIPER GREGORY F', 'HUMPHREY GENE E', 'UMANOFF ADAM S', 'BLACHMAN JEREMY M', 'SUNDE MARTIN', 'GIBBS DANA R', 'LOWRY CHARLES P', 'COLWELL WESLEY', 'MULLER MARK S', 'JACKSON CHARLENE R', 'WESTFAHL RICHARD K', 'WALTERS GARETH W', 'WALLS JR ROBERT H', 'KITCHEN LOUISE', 'CHAN RONNIE', 'BELFER ROBERT', 'SHANKMAN JEFFREY A', 'WODRASKA JOHN', 'BERGSIEKER RICHARD P', 'URQUHART JOHN A', 'BIBI PHILIPPE A', 'RIEKER PAULA H', 'WHALEY DAVID A', 'BECK SALLY W', 'HAUG DAVID L', 'ECHOLS JOHN B', 'MENDELSOHN JOHN', 'HICKERSON GARY J', 'CLINE KENNETH W', 'LEWIS RICHARD', 'HAYES ROBERT E', 'MCCARTY DANNY J', 'KOPPER MICHAEL J', 'LEFF DANIEL P', 'LAVORATO JOHN J', 'BERBERIAN DAVID', 'DETMERING TIM

I can see from this brief exploration that there are 146 exectives in the dataset, 18 identified POIs, and 22 features, for a total of 3088 observations. There are also a number of missing values. I'll investigate those in the next section.

### EDA and Outlier Removal

In [5]:
#change dataset to pandas dataframe
df = pandas.DataFrame.from_records(list(data_dict.values()))
employees = pandas.Series(list(data_dict.keys()))

#count number of NA values
df.replace(to_replace='NaN', value=numpy.nan, inplace=True)
print "Number of NaN values for each feature:"
print df.isnull().sum()
print "Shape of the dataframe: ", df.shape

Number of NaN values for each feature:
bonus                         64
deferral_payments            107
deferred_income               97
director_fees                129
email_address                 35
exercised_stock_options       44
expenses                      51
from_messages                 60
from_poi_to_this_person       60
from_this_person_to_poi       60
loan_advances                142
long_term_incentive           80
other                         53
poi                            0
restricted_stock              36
restricted_stock_deferred    128
salary                        51
shared_receipt_with_poi       60
to_messages                   60
total_payments                21
total_stock_value             20
dtype: int64
Shape of the dataframe:  (146, 21)


There are quite a lot of NaN values for some of the features. Particularly loan advances, director fees, restricted stock deferred, and deferral payments. However, when I look at the data schema provided (enron61702insiderpay.pdf), I see that these "missing values" are actually 0, so I will convert them to 0.

In [23]:
#replace missing values with 0
df.replace(to_replace=numpy.nan, value=0, inplace=True)
#drop email address column
del df['email_address']

df.describe()

['bonus', 'deferral_payments', 'deferred_income', 'director_fees', 'exercised_stock_options', 'expenses', 'from_messages', 'from_poi_to_this_person', 'from_this_person_to_poi', 'loan_advances', 'long_term_incentive', 'other', 'poi', 'restricted_stock', 'restricted_stock_deferred', 'salary', 'shared_receipt_with_poi', 'to_messages', 'total_payments', 'total_stock_value', 'poi_email_ratio', 'payment_ratio']


bonus  deferral_payments  deferred_income  director_fees  \
count  1.420000e+02       1.420000e+02     1.420000e+02     142.000000   
mean   6.819973e+05       2.259394e+05    -1.964112e+05   10097.753521   
std    1.240462e+06       7.586983e+05     6.098517e+05   31505.310400   
min    0.000000e+00       0.000000e+00    -3.504386e+06       0.000000   
25%    0.000000e+00       0.000000e+00    -3.792600e+04       0.000000   
50%    3.000000e+05       0.000000e+00     0.000000e+00       0.000000   
75%    8.000000e+05       9.684500e+03     0.000000e+00       0.000000   
max    8.000000e+06       6.426990e+06     0.000000e+00  137864.000000   

       exercised_stock_options       expenses  from_messages  \
count             1.420000e+02     142.000000     142.000000   
mean              2.099202e+06   35869.359155     367.823944   
std               4.825043e+06   45435.170550    1460.462179   
min               0.000000e+00       0.000000       0.000000   
25%               0.000000e+00       0.000000       0.000000   
50%               6.082935e+05   21937.000000      17.500000   
75%               1.714221e+06   53740.750000      51.250000   
max               3.434838e+07  228763.000000   14368.000000   

       from_poi_to_this_person  from_this_person_to_poi  loan_advances  \
count               142.000000               142.000000   1.420000e+02   
mean                 39.302817                24.971831   5.910211e+05   
std                  74.657135                80.287775   6.842133e+06   
min                   0.000000                 0.000000   0.000000e+00   
25%                   0.000000                 0.000000   0.000000e+00   
50%                   5.500000                 0.000000   0.000000e+00   
75%                  41.750000                14.000000   0.000000e+00   
max                 528.000000               609.000000   8.152500e+07   

           ...               other  restricted_stock  \
count      ...        1.420000e+02      1.420000e+02   
mean       ...        2.988883e+05      8.794379e+05   
std        ...        1.138775e+06      2.028670e+06   
min        ...        0.000000e+00     -2.604490e+06   
25%        ...        0.000000e+00      3.536825e+04   
50%        ...        9.190000e+02      3.619780e+05   
75%        ...        1.498310e+05      8.145280e+05   
max        ...        1.035973e+07      1.476169e+07   

       restricted_stock_deferred        salary  shared_receipt_with_poi  \
count               1.420000e+02  1.420000e+02               142.000000   
mean                7.414144e+04  1.863114e+05               712.345070   
std                 1.311168e+06  1.977470e+05              1081.732338   
min                -1.787380e+06  0.000000e+00                 0.000000   
25%                 0.000000e+00  0.000000e+00                 0.000000   
50%                 0.000000e+00  2.105960e+05               115.500000   
75%                 0.000000e+00  2.708505e+05              1001.250000   
max                 1.545629e+07  1.111258e+06              5521.000000   

        to_messages  total_payments  total_stock_value  poi_email_ratio  \
count    142.000000    1.420000e+02       1.420000e+02       142.000000   
mean    1248.338028    2.280176e+06       2.943935e+06         0.028894   
std     2250.905721    8.907260e+06       6.225651e+06         0.042994   
min        0.000000    0.000000e+00       0.000000e+00         0.000000   
25%        0.000000    8.839225e+04       2.534955e+05         0.000000   
50%      347.500000    9.413595e+05       9.659550e+05         0.009109   
75%     1655.000000    1.968287e+06       2.319991e+06         0.043673   
max    15149.000000    1.035598e+08       4.911008e+07         0.224352   

       payment_ratio  
count     142.000000  
mean        0.460213  
std         0.342553  
min         0.000000  
25%         0.124135  
50%         0.446557  
75%         0.719112  
max         1.000000  

[8 rows x 21 columns]

In [7]:
#pairplot to visualize feature distributions
def splom_viz(df, labels=None):
    ax = sns.pairplot(df, hue="poi", diag_kind='kde', size=2, vars=['poi','salary', 'total_payments', 'bonus', 
                 'deferred_income', 'total_stock_value', 'expenses', 'exercised_stock_options', 'other', 'long_term_incentive', 
                 'restricted_stock', 'to_messages', 'from_poi_to_this_person', 'from_messages',
                 'from_this_person_to_poi', 'shared_receipt_with_poi'])
    plt.show()

#splom_viz(df)

Looking at this data, there definitely appears to be outliers. If I look at the data dictionary keys again, I see that there are two that are not names: Total and travel agency in the park. I'll remove these then look again at the pairplot.

#### Outlier Removal

In [31]:
#outlier removal
df= df.drop(df.index[[data_dict.keys().index("TOTAL"), data_dict.keys().index("THE TRAVEL AGENCY IN THE PARK")]])
df.describe()

#pairplot to visualize distributions and correllations
splom_viz(df)

After these outliers are removed, I can see on the pairplot that the most of the remainder of the outliers are classified as POI, so these "outliers" are in fact real data. The exception to this are the features 'from_poi_to_this_person', 'from_this_person_to_poi', and 'from_messages'. Looking at the statistics above, we see that the max 'from_messages' is 14368, which is one order of magnitude higher than the 75%. The same is true for the outliers in the other two categories.  Who are these people?

In [9]:
#identify keys for potential outliers
for key, value in data_dict.items():
    if value['from_poi_to_this_person'] != 'NaN' and value['from_poi_to_this_person'] > 500: 
        print "Max from_poi_to_this_person: ", key

for key, value in data_dict.items():
    if value['from_this_person_to_poi'] != 'NaN' and value['from_this_person_to_poi'] > 500: 
        print "Max from_this_person_to_poi: ", key
        
for key, value in data_dict.items():
    if value['from_messages'] != 'NaN' and value['from_messages'] > 14000: 
        print "Max from_messages: ", key

Max from_poi_to_this_person:  LAVORATO JOHN J
Max from_this_person_to_poi:  DELAINEY DAVID W
Max from_messages:  KAMINSKI WINCENTY J


Since these keys are all different people, I will keep the email data and assume it is real. Finally, looking at the pairplots and statistics for each feature, there are negative values for deferred_income, defferal_payments, restricted_stock, and restricted_stock deferred. Are these outliers, real data, or errors in the dataset?

#### Checking financial features

One final check we can do is make sure there aren't any mistakes in the financial data. In the data schema, we can see that two features: total_payments and total_stock_value are linear combinations of the other financial features. If the negative values observed above are real, then total_payments and total_stock_value should equal the sum of the values these features. If there is an error in the dataset, then these values will not be equal.

In [10]:
#Checking total pay
df_totalPay = pandas.DataFrame()
df_totalPay['name'] = data_dict.keys()
df_totalPay['total'] = (
    df['bonus'] + 
    df['director_fees'] +
    df['deferral_payments'] +
    df['deferred_income'] +
    df['loan_advances'] +
    df['long_term_incentive'] +
    df['expenses'] +
    df['other'] +
    df['salary']
)
df_totalPay['total_payments'] = df['total_payments']
df_totalPay['equals?'] = (df_totalPay['total'] == df_totalPay['total_payments'])
df_totalPay['poi'] = df['poi']
print numpy.sum(df_totalPay['equals?']), " out of ", len(df_totalPay)
print "Summed Totals Different than Total Payments"
print df_totalPay[df_totalPay['equals?'] == False]

142  out of  146
Summed Totals Different than Total Payments
                              name     total  total_payments equals?    poi
24                   BELFER ROBERT  -99215.0        102500.0   False  False
101  THE TRAVEL AGENCY IN THE PARK       NaN             NaN   False    NaN
104                          TOTAL       NaN             NaN   False    NaN
118               BHATNAGAR SANJAY  275728.0      15456290.0   False  False


In [11]:
#Checking total stock value
df_totalStock = pandas.DataFrame()
df_totalStock['name'] = data_dict.keys()
df_totalStock['total'] = (
    df['restricted_stock'] +
    df['exercised_stock_options'] +
    df['restricted_stock_deferred']
)
df_totalStock['total_stock_value'] = df['total_stock_value']
df_totalStock['equals?'] = (df_totalStock['total'] ==df_totalStock['total_stock_value'])
df_totalStock['poi'] = df['poi']
print numpy.sum(df_totalPay['equals?']), " out of ", len(df_totalPay)
print "Summed Totals Different from Total Stock Value"
print df_totalStock[df_totalStock['equals?'] == False]

142  out of  146
Summed Totals Different from Total Stock Value
                              name       total  total_stock_value equals?  \
24                   BELFER ROBERT     47378.0           -44093.0   False   
101  THE TRAVEL AGENCY IN THE PARK         NaN                NaN   False   
104                          TOTAL         NaN                NaN   False   
118               BHATNAGAR SANJAY  15456290.0                0.0   False   

       poi  
24   False  
101    NaN  
104    NaN  
118  False  


In [12]:
#check 
print data_dict["BELFER ROBERT"]
print"/n"
print data_dict["BHATNAGAR SANJAY"]

{'salary': 'NaN', 'to_messages': 'NaN', 'deferral_payments': -102500, 'total_payments': 102500, 'exercised_stock_options': 3285, 'bonus': 'NaN', 'restricted_stock': 'NaN', 'shared_receipt_with_poi': 'NaN', 'restricted_stock_deferred': 44093, 'total_stock_value': -44093, 'expenses': 'NaN', 'loan_advances': 'NaN', 'from_messages': 'NaN', 'other': 'NaN', 'from_this_person_to_poi': 'NaN', 'poi': False, 'director_fees': 3285, 'deferred_income': 'NaN', 'long_term_incentive': 'NaN', 'email_address': 'NaN', 'from_poi_to_this_person': 'NaN'}
/n
{'salary': 'NaN', 'to_messages': 523, 'deferral_payments': 'NaN', 'total_payments': 15456290, 'exercised_stock_options': 2604490, 'bonus': 'NaN', 'restricted_stock': -2604490, 'shared_receipt_with_poi': 463, 'restricted_stock_deferred': 15456290, 'total_stock_value': 'NaN', 'expenses': 'NaN', 'loan_advances': 'NaN', 'from_messages': 29, 'other': 137864, 'from_this_person_to_poi': 1, 'poi': False, 'director_fees': 137864, 'deferred_income': 'NaN', 'long_t

It appears that the values for these two employees are incorrect when I compare the data schema to the data dictionary. I will drop these two individuals from the dataframe.

In [13]:
#remove two outliers
df = df.drop(df.index[[data_dict.keys().index("BELFER ROBERT"), data_dict.keys().index("BHATNAGAR SANJAY")]])
df.describe()

bonus  deferral_payments  deferred_income  director_fees  \
count  1.420000e+02       1.420000e+02     1.420000e+02     142.000000   
mean   6.819973e+05       2.259394e+05    -1.964112e+05   10097.753521   
std    1.240462e+06       7.586983e+05     6.098517e+05   31505.310400   
min    0.000000e+00       0.000000e+00    -3.504386e+06       0.000000   
25%    0.000000e+00       0.000000e+00    -3.792600e+04       0.000000   
50%    3.000000e+05       0.000000e+00     0.000000e+00       0.000000   
75%    8.000000e+05       9.684500e+03     0.000000e+00       0.000000   
max    8.000000e+06       6.426990e+06     0.000000e+00  137864.000000   

       exercised_stock_options       expenses  from_messages  \
count             1.420000e+02     142.000000     142.000000   
mean              2.099202e+06   35869.359155     367.823944   
std               4.825043e+06   45435.170550    1460.462179   
min               0.000000e+00       0.000000       0.000000   
25%               0.000000e+00       0.000000       0.000000   
50%               6.082935e+05   21937.000000      17.500000   
75%               1.714221e+06   53740.750000      51.250000   
max               3.434838e+07  228763.000000   14368.000000   

       from_poi_to_this_person  from_this_person_to_poi  loan_advances  \
count               142.000000               142.000000   1.420000e+02   
mean                 39.302817                24.971831   5.910211e+05   
std                  74.657135                80.287775   6.842133e+06   
min                   0.000000                 0.000000   0.000000e+00   
25%                   0.000000                 0.000000   0.000000e+00   
50%                   5.500000                 0.000000   0.000000e+00   
75%                  41.750000                14.000000   0.000000e+00   
max                 528.000000               609.000000   8.152500e+07   

       long_term_incentive         other  restricted_stock  \
count         1.420000e+02  1.420000e+02      1.420000e+02   
mean          3.395572e+05  2.988883e+05      8.794379e+05   
std           6.914468e+05  1.138775e+06      2.028670e+06   
min           0.000000e+00  0.000000e+00     -2.604490e+06   
25%           0.000000e+00  0.000000e+00      3.536825e+04   
50%           0.000000e+00  9.190000e+02      3.619780e+05   
75%           3.750648e+05  1.498310e+05      8.145280e+05   
max           5.145434e+06  1.035973e+07      1.476169e+07   

       restricted_stock_deferred        salary  shared_receipt_with_poi  \
count               1.420000e+02  1.420000e+02               142.000000   
mean                7.414144e+04  1.863114e+05               712.345070   
std                 1.311168e+06  1.977470e+05              1081.732338   
min                -1.787380e+06  0.000000e+00                 0.000000   
25%                 0.000000e+00  0.000000e+00                 0.000000   
50%                 0.000000e+00  2.105960e+05               115.500000   
75%                 0.000000e+00  2.708505e+05              1001.250000   
max                 1.545629e+07  1.111258e+06              5521.000000   

        to_messages  total_payments  total_stock_value  
count    142.000000    1.420000e+02       1.420000e+02  
mean    1248.338028    2.280176e+06       2.943935e+06  
std     2250.905721    8.907260e+06       6.225651e+06  
min        0.000000    0.000000e+00       0.000000e+00  
25%        0.000000    8.839225e+04       2.534955e+05  
50%      347.500000    9.413595e+05       9.659550e+05  
75%     1655.000000    1.968287e+06       2.319991e+06  
max    15149.000000    1.035598e+08       4.911008e+07

### Feature Selection

#### Create New Feature

Before I look at selecting the features for my classifier, I'll first create some new features.  I want to see if the ratio of total payments/total compensation is an indicator of POIs. I'll also look at the ratio of emails involving POIs to all other emails as a new feature. Then, I'll let feature selection decide if I use these features in the model.

In [14]:
#create new feature: ratio of messages involving POI/total
df['poi_email_ratio'] = (df['from_poi_to_this_person'] + df['from_this_person_to_poi']) / (df['from_messages'] + df['to_messages'])
df['poi_email_ratio'].replace(to_replace='NaN', value=0, inplace=True)
#new feature payments/total compensation ratio
df['payment_ratio'] = df['total_payments']/(df['total_stock_value'] + df['total_payments'])
df.describe()

bonus  deferral_payments  deferred_income  director_fees  \
count  1.420000e+02       1.420000e+02     1.420000e+02     142.000000   
mean   6.819973e+05       2.259394e+05    -1.964112e+05   10097.753521   
std    1.240462e+06       7.586983e+05     6.098517e+05   31505.310400   
min    0.000000e+00       0.000000e+00    -3.504386e+06       0.000000   
25%    0.000000e+00       0.000000e+00    -3.792600e+04       0.000000   
50%    3.000000e+05       0.000000e+00     0.000000e+00       0.000000   
75%    8.000000e+05       9.684500e+03     0.000000e+00       0.000000   
max    8.000000e+06       6.426990e+06     0.000000e+00  137864.000000   

       exercised_stock_options       expenses  from_messages  \
count             1.420000e+02     142.000000     142.000000   
mean              2.099202e+06   35869.359155     367.823944   
std               4.825043e+06   45435.170550    1460.462179   
min               0.000000e+00       0.000000       0.000000   
25%               0.000000e+00       0.000000       0.000000   
50%               6.082935e+05   21937.000000      17.500000   
75%               1.714221e+06   53740.750000      51.250000   
max               3.434838e+07  228763.000000   14368.000000   

       from_poi_to_this_person  from_this_person_to_poi  loan_advances  \
count               142.000000               142.000000   1.420000e+02   
mean                 39.302817                24.971831   5.910211e+05   
std                  74.657135                80.287775   6.842133e+06   
min                   0.000000                 0.000000   0.000000e+00   
25%                   0.000000                 0.000000   0.000000e+00   
50%                   5.500000                 0.000000   0.000000e+00   
75%                  41.750000                14.000000   0.000000e+00   
max                 528.000000               609.000000   8.152500e+07   

           ...               other  restricted_stock  \
count      ...        1.420000e+02      1.420000e+02   
mean       ...        2.988883e+05      8.794379e+05   
std        ...        1.138775e+06      2.028670e+06   
min        ...        0.000000e+00     -2.604490e+06   
25%        ...        0.000000e+00      3.536825e+04   
50%        ...        9.190000e+02      3.619780e+05   
75%        ...        1.498310e+05      8.145280e+05   
max        ...        1.035973e+07      1.476169e+07   

       restricted_stock_deferred        salary  shared_receipt_with_poi  \
count               1.420000e+02  1.420000e+02               142.000000   
mean                7.414144e+04  1.863114e+05               712.345070   
std                 1.311168e+06  1.977470e+05              1081.732338   
min                -1.787380e+06  0.000000e+00                 0.000000   
25%                 0.000000e+00  0.000000e+00                 0.000000   
50%                 0.000000e+00  2.105960e+05               115.500000   
75%                 0.000000e+00  2.708505e+05              1001.250000   
max                 1.545629e+07  1.111258e+06              5521.000000   

        to_messages  total_payments  total_stock_value  poi_email_ratio  \
count    142.000000    1.420000e+02       1.420000e+02       142.000000   
mean    1248.338028    2.280176e+06       2.943935e+06         0.028894   
std     2250.905721    8.907260e+06       6.225651e+06         0.042994   
min        0.000000    0.000000e+00       0.000000e+00         0.000000   
25%        0.000000    8.839225e+04       2.534955e+05         0.000000   
50%      347.500000    9.413595e+05       9.659550e+05         0.009109   
75%     1655.000000    1.968287e+06       2.319991e+06         0.043673   
max    15149.000000    1.035598e+08       4.911008e+07         0.224352   

       payment_ratio  
count     139.000000  
mean        0.470146  
std         0.339396  
min         0.000000  
25%         0.160776  
50%         0.452582  
75%         0.724538  
max         1.000000  

[8 rows x 21 columns]

#### Visualize Feature Separation/Correlation

In [30]:
#radviz plot
from pandas.tools.plotting import radviz
radviz(df, 'poi')
plt.show()

#correlation table
print "Pearson correlation between features:"
print df.corr(method='pearson')

Pearson correlation between features:
                              bonus  deferral_payments  deferred_income  \
bonus                      1.000000           0.173090        -0.330798   
deferral_payments          0.173090           1.000000        -0.542689   
deferred_income           -0.330798          -0.542689         1.000000   
director_fees             -0.177464          -0.096124         0.077079   
exercised_stock_options    0.415674           0.116331        -0.267636   
expenses                   0.243604          -0.024248        -0.038320   
from_messages              0.173714           0.028069        -0.013916   
from_poi_to_this_person    0.640544           0.215175        -0.193366   
from_this_person_to_poi    0.448018           0.000789        -0.003598   
loan_advances              0.432479           0.014256        -0.024998   
long_term_incentive        0.497020           0.118576        -0.295811   
other                      0.383004           0.368232        

In this radviz plot, there is some separation between non-POIs and POIs. With the more predictive features for non-POIs being deferral payments, deferred income, and bonus. This is a multi-dimensional dataset though, and radviz is not great at visualizing high dimensional data. The correlation table also shows that there is no strong correlation between POI and other features. Because of this, I'll use PCA and/or Select K best for feature selection.

#### Feature Scaling and Selection

In [50]:
#convert pandas df to pickled dictionary
#drop rows in index corresponding to df
employees = employees = pandas.Series(list(data_dict.keys()))
employees = employees.drop(employees.index[[24, 101, 103, 104, 107, 120]])

#create new feature list
new_features_list = df.columns.values

# set the index of df to be the employees series:
df.set_index(employees, inplace=True)

# create a dictionary from the dataframe
df_dict = df.to_dict('index')

# Store to my_dataset for easy export below.
my_dataset = df_dict
my_feature_list = ['poi','bonus', 'deferral_payments', 'deferred_income', 'director_fees',
 'exercised_stock_options', 'expenses', 'from_messages',
 'from_poi_to_this_person', 'from_this_person_to_poi', 'loan_advances',
 'long_term_incentive', 'other', 'restricted_stock',
 'restricted_stock_deferred', 'salary', 'shared_receipt_with_poi',
 'to_messages', 'total_payments', 'total_stock_value', 'poi_email_ratio',
 'payment_ratio']
# Extract features and labels from dataset for local testing
data = featureFormat(my_dataset, my_feature_list, sort_keys = True)
labels, features = targetFeatureSplit(data)

Since there is a large range in the data from binary data (poi), to 10s-100s of emails, to a salary range averaging 10^6, I'm going to standardize all the data using sklearn's Standard Scaler. This converts all the features into a normal distribution with a mean of 0 and a standard deviation of 1.  I chose to use standardization instead of normalization (used in sklearn's MinMaxScaler) because normalization will suppress the effects of outliers, whereas standardization won't. I've seen during the data exploration above that there are still significant outliers in the data and standardization will keep information about these outliers instead of suppressing them.

In [57]:
#Use scaler to standardize data
scaler = preprocessing.StandardScaler()
features = scaler.fit_transform(features)

Before building the classifiers, I'll go ahead and use SelectKBest to choose my features. I'll do this before splitting into training and testing sets since I'll be using StratifiedShuffleSplit to separate my data. If I did it after splitting my data, then each time I run the classifier I may get different results for feature selection.

In [62]:
# K-best features
k_best = SelectKBest(f_classif, k=10)
k_best.fit(features, labels)

new_features = k_best.transform(features)

#print features and importances
results_list = zip(k_best.get_support(), my_features[1:], k_best.scores_)
results_list = sorted(results_list, key=lambda x: x[2], reverse=True)
print "K-best features:", results_list

K-best features: [(True, 'total_stock_value', 25.231232446123194), (True, 'exercised_stock_options', 23.93469840457637), (True, 'bonus', 20.146822586083687), (True, 'salary', 18.62161412135108), (True, 'restricted_stock', 16.001429788534193), (True, 'deferred_income', 10.911025402562961), (True, 'long_term_incentive', 9.4791364272618868), (True, 'shared_receipt_with_poi', 9.1304349912039857), (True, 'total_payments', 8.468703818662398), (True, 'loan_advances', 6.9493725284790617), (False, 'expenses', 5.9775270509968053), (False, 'from_poi_to_this_person', 5.1843501345700904), (False, 'poi_email_ratio', 5.0050854157011306), (False, 'other', 4.0946439501793979), (False, 'from_this_person_to_poi', 3.1770230797679671), (False, 'to_messages', 2.4180979910905469), (False, 'director_fees', 2.1933154116755791), (False, 'deferral_payments', 0.26155708920673809), (False, 'from_messages', 0.095591541756847059), (False, 'restricted_stock_deferred', 0.067080895794596071), (False, 'payment_ratio', 0

In [55]:
#split into training and testing data
features_train, features_test, labels_train, labels_test = cross_validation.train_test_split(features, labels, test_size=0.1, random_state=42)

### Create/tune classifiers

In [86]:
#SVM in hyperopt
from sklearn.svm import SVC

def hyperopt_train_test(params):
    features_train_ = features_train[:]

    clf = SVC(**params)
    return cross_val_score(clf, features_train_, labels_train).mean()

space4svm = {
    'C': hp.uniform('C', 0, 20),
    'kernel': hp.choice('kernel', ['linear', 'sigmoid', 'poly', 'rbf']),
    'gamma': hp.uniform('gamma', 0, 20)
}

def f(params):
    acc = hyperopt_train_test(params)
    return {'loss': -acc, 'status': STATUS_OK}

trials = Trials()
best = fmin(f, space4svm, algo=tpe.suggest, max_evals=100, trials=trials)
print 'best:'
print best

clf = SVC(kernel='linear', C=3.182443373076033, gamma= 2.7858303946931793)
clf.fit(features_train, labels_train)
pred = clf.predict(features_test)
accuracy = accuracy_score(pred,labels_test)
print 'accuracy = ', accuracy_score(pred,labels_test)
print 'precision = ', precision_score(labels_test,pred)
print 'recall = ', recall_score(labels_test,pred)

best:
{'kernel': 1, 'C': 3.1824433730760333, 'gamma': 2.7858303946931793}
accuracy =  0.928571428571
precision =  0.0
recall =  0.0


C:\Users\JoAnna\Anaconda2\lib\site-packages\sklearn\metrics\classification.py:1074: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)


In [ ]:
### Task 4: Try a varity of classifiers
### Please name your classifier clf for easy export below.
### Note that if you want to do PCA or other multi-stage operations,
### you'll need to use Pipelines. For more info:
### http://scikit-learn.org/stable/modules/pipeline.html

# Provided to give you a starting point. Try a variety of classifiers.
from sklearn.naive_bayes import GaussianNB
clf = GaussianNB()

In [ ]:
### Task 5: Tune your classifier to achieve better than .3 precision and recall 
### using our testing script. Check the tester.py script in the final project
### folder for details on the evaluation method, especially the test_classifier
### function. Because of the small size of the dataset, the script uses
### stratified shuffle split cross validation. For more info: 
### http://scikit-learn.org/stable/modules/generated/sklearn.cross_validation.StratifiedShuffleSplit.html

# Example starting point. Try investigating other evaluation techniques!
from sklearn.cross_validation import train_test_split
features_train, features_test, labels_train, labels_test = \
    train_test_split(features, labels, test_size=0.3, random_state=42)

In [ ]:
### Task 6: Dump your classifier, dataset, and features_list so anyone can
### check your results. You do not need to change anything below, but make sure
### that the version of poi_id.py that you submit can be run on its own and
### generates the necessary .pkl files for validating your results.

dump_classifier_and_data(clf, my_dataset, my_feature_list)